signify.c

```diff
@@ -547,6 +548,35 @@ verifysimple(const char *pubkeyfile, const char *msgfile, const char *sigfile,
 	free(msg);
 }
 
+static int
+fingerprint(const char *seckeyfile, const char *pubkeyfile, const char *sigfile)
+{
+	struct sig sig;
+	struct pubkey pubkey;
+	struct enckey enckey;
+	uint8_t *fp;
+
+	if (seckeyfile) {
+		readb64file(seckeyfile, &enckey, sizeof(enckey), NULL);
+		fp = enckey.keynum;
+	} else if (pubkeyfile) {
+		readb64file(pubkeyfile, &pubkey, sizeof(pubkey), NULL);
+		fp = pubkey.keynum;
+	} else if (sigfile) {
+		readb64file(sigfile, &sig, sizeof(sig), NULL);
```

Comment:
You are reading base64-decoded data directly into a struct on the stack without input validation, which is just asking for a buffer overflow.